### PR TITLE
feat: update mobile-ui

### PR DESCRIPTION
**Pull request by JeikZim**
Changes:
 - new mobile SubNavigation
 - new mobile MiniToc;
 - new share button;
 - new mobile Sidebar with Toc and main menu;
 - new language and theme settings;
 - search has been moved to the header
 - updated dropdown in the mobile menu
 - changed the styles of the Feedback block
 - changed the styles of TocNavPanel
 - fixed a bug with search on mobile devices

### CODE REVIEW

File: src/components/DocPage/DocPage.tsx [hunk 1]
```diff
@@ -152,12 +166,20 @@ class DocPage extends React.Component<DocPageInnerProps, DocPageState> {
         } = this.props;
 
         const hideMiniToc = !this.showMiniToc;
+        const hideBurger = typeof headerHeight !== 'undefined' && headerHeight !== 0;
```
Comment (martyanovandrey) [inline src/components/DocPage/DocPage.tsx:169]: probably undefined check is not needed here

File: src/components/Feedback/controls/LikeControl.tsx [hunk 2]
```diff
@@ -22,42 +24,49 @@ type LikeControlProps = {
...
-    forwardRef<HTMLButtonElement, LikeControlProps>(({isLiked, view, onClick}, ref) => {
-        const {t} = useTranslation('feedback');
-        const {isVerticalView, popupPosition, controlClassName} = useContext(ControlsLayoutContext);
-        const tooltipText = isLiked ? t('cancel-like-text') : t('like-text');
+    forwardRef<HTMLButtonElement, LikeControlProps>(
+        ({isLiked, isPopupVisible, view, onClick}, ref) => {
+            const {t} = useTranslation('feedback');
+            const {isVerticalView, popupPosition, controlClassName} =
+                useContext(ControlsLayoutContext);
+            const tooltipText = isLiked ? t('cancel-like-text') : t('like-text');
```
Comment (martyanovandrey) [inline src/components/Feedback/controls/LikeControl.tsx:32]: if the value is used only once, it is not necessary to put it in a separate variable

File: src/components/MobileControls/MobileControl/MobileControl.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControl/MobileControl.tsx:1]: /* eslint-disable @typescript-eslint/no-explicit-any */ ?

File: src/components/MobileControls/MobileControl/MobileControl.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            [item, displayItems],
+        );
+
+        const [sheetIsVisible, setSheetIsVisible] = useState(false);
+
+        const onSheetOpen = () => setSheetIsVisible(true);
+        const onSheetClose = () => setSheetIsVisible(false);
+        const onItemClick = (value: any) => (onChangeValue ? onChangeValue(value) : null);
+
+        const label = (
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControl/MobileControl.tsx:37]: why is this element defined separately ?

File: src/components/MobileControls/MobileControlSheet/MobileControlSheet.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import {List, Sheet} from '@gravity-ui/uikit';
+import block from 'bem-cn-lite';
+
+import {ListItem} from '../../../models';
+
+import './MobileControlSheet.scss';
+
+const b = block('dc-mobile-control-sheet');
+
+export interface MobileControlSheetProps {
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControlSheet/MobileControlSheet.tsx:11]: It would be better if the type definitions were after or before, and not between, the variable declarations.

File: src/components/navigation/Sidebar/Sidebar.scss [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+@import '../../../styles/mixins.scss';
+
+.dc-sidebar {
+    $transitionTime: 0.4s;
+
+    @include mobile-tablet-only();
+
+    position: fixed;
+    top: var(--dc-header-height, #{$headerHeight});
+    // TODO: delete
```
Comment (martyanovandrey) [inline src/components/navigation/Sidebar/Sidebar.scss:11]: there are a lot of unimplemented tudos and commented code in this PR, it is worth fixing

File: src/components/MobileControls/MobileControls.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControls.tsx:1]: eslint-disable is not the right way to solve ts errors

File: src/components/SubNavigation/SubNavigation.scss [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        flex: 1 0 0;
+
+        width: 100%;
+        padding: 0;
+        padding-left: 4px;
+        gap: 4px;
+
+        border-radius: var(--g-border-radius-xs, 3px);
+        border: none;
+        border-left: 1px solid var(--g-color-line-generic, rgba(0, 0, 0, 0.1019607843));
```
Comment (martyanovandrey) [inline src/components/SubNavigation/SubNavigation.scss:68]: what is `rgba(0, 0, 0, 0.1019607843)` ?

File: src/components/SubNavigation/hooks/useShortenedTitle.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    2 * widthConfig.gap;
+
+const speacialChars = new Set([',', '.', '?', '!', '-', '_', ':', ';', '(', '[', '{']);
+
+const getAnotherContentWidth = (hideBurger: boolean) =>
+    hideBurger ? widthWithoutBurger : widthWithBurger;
+// the values are calculated empirically
+const getSymbolSizeQuotient = (hideBurger: boolean) => (hideBurger ? 1 / 7.0 : 1 / 6.5);
+
+// ",...", "?...", "!...", "-...", "_...", "(...", ":..." - looks bad
```
Comment (martyanovandrey) [inline src/components/SubNavigation/hooks/useShortenedTitle.ts:34]: what is the result of this hook? why didn't css work here for visual string trimming ?

File: src/components/navigation/CustomNavigation/CustomNavigation.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+import block from 'bem-cn-lite';
+
+import {MobileControlsProps} from '../../MobileControls/MobileControls';
+import Sidebar from '../Sidebar/Sidebar';
+import SidebarContent, {NavigationTocProps} from '../SidebarContent/SidebarContent';
+import {useItemsWithCustomDropdowns} from '../hooks';
+
+import './CustomNavigation.scss';
+
+const CLASS_NAME = 'pc-navigation';
```
Comment (martyanovandrey) [inline src/components/navigation/CustomNavigation/CustomNavigation.tsx:21]: we reuse navigation styles here or make our own custom one ?

File: src/i18n/he.json [hunk 11]
```diff
@@ -45,6 +45,9 @@
     "lang-control-text-ru": "Русский",
     "lang-control-text-en": "English"
   },
+  "mobile-menu": {
+    "label_to-main-menu": "חזור לתפריט הראשי"
```
Comment (martyanovandrey) [inline src/i18n/he.json:49]: need to check how the new mobile interface works for rtl documentation

File: src/i18n/fr.json [hunk 12]
```diff
@@ -45,6 +45,9 @@
     "lang-control-text-ru": "Русский",
     "lang-control-text-en": "English"
   },
+  "mobile-menu": {
+    "label_to-main-menu": "Retour au menu principal"
```
Comment (martyanovandrey) [inline src/i18n/fr.json:49]: Translations should be added only for Russian and English, other languages ​​via tanker

**Review by martyanovandrey — CHANGES_REQUESTED** [review]

File: src/components/DocPage/DocPage.tsx (hunk 1 above)
Comment (JeikZim) [reply src/components/DocPage/DocPage.tsx]: A variable can be undefined, so before checking its numeric value, we need to make sure that it is not undefined. But I will replace the headerHeight !== 0 on headerHeight > 0, because the height cannot be less than zero

File: src/components/MobileControls/MobileControl/MobileControl.tsx (hunk 4 above)
Comment (JeikZim) [reply src/components/MobileControls/MobileControl/MobileControl.tsx]: At that moment, it seemed to me the best solution in terms of readability, but now I have changed this part of the code

File: src/components/Feedback/controls/LikeControl.tsx (hunk 2 above)
Comment (JeikZim) [reply src/components/Feedback/controls/LikeControl.tsx:32]: Sorry, I don't understand what part of the code you're talking about, because what you highlighted I didn't write, it was written outside my pull-request

File: src/i18n/he.json (hunk 11 above)
Comment (JeikZim) [reply src/i18n/he.json]: Thanks, I checked and found a few bugs. I've already fixed it

File: src/components/navigation/CustomNavigation/CustomNavigation.tsx (hunk 10 above)
Comment (JeikZim) [reply src/components/navigation/CustomNavigation/CustomNavigation.tsx:21]: Most of our styles, and some of the page-constructor styles, use the .pc-navigation class. Therefore, to avoid problems and just repeat the styles in the previous form, I used this class name

File: src/components/SubNavigation/hooks/useShortenedTitle.ts (hunk 9 above)
Comment (JeikZim) [reply src/components/SubNavigation/hooks/useShortenedTitle.ts]: I made a big mistake, and I didn't know that this task could be solved with CSS alone. Now I have used CSS, and now this hook is not needed

File: src/components/SubNavigation/SubNavigation.scss (hunk 8 above)
Comment (JeikZim) [reply src/components/SubNavigation/SubNavigation.scss:72]: This is the exact value of the border color taken from Figma

File: src/components/MobileControls/MobileControl/MobileControl.tsx (hunk 3 above)
Comment (JeikZim) [reply src/components/MobileControls/MobileControl/MobileControl.tsx:1]: The onChangeValue and onItemClick functions accept the "value" argument, which is of any type, so I used any. 
But now I have replaced 
/* eslint-disable @typescript-eslint/no-explicit-any */
with 
// eslint-disable-next-line @typescript-eslint/no-explicit-any

File: src/components/Feedback/controls/LikeControl.tsx [hunk 13]
```diff
@@ -22,42 +24,49 @@ type LikeControlProps = {
...
-                    {t<string>('button-like-text')}
-                </Button>
+                    className={b('control', {view}, controlClassName)}
+                    isVerticalView={isVerticalView}
+                    tooltipText={tooltipText}
+                    ref={ref}
+                    icon={Icon}
+                    popupPosition={popupPosition}
+                    buttonExtraProps={{
+                        'aria-expanded': isPopupVisible,
```
Comment (martyanovandrey) [inline src/components/Feedback/controls/LikeControl.tsx:64]: Try to rebase on master. There should be no 'expanded' i deleted them in this commit.

https://github.com/diplodoc-platform/components/commit/d74fbd4f899251afcae000b9fca31aec105a1470

**Review by martyanovandrey — CHANGES_REQUESTED** [review]

File: src/components/MiniToc/MiniToc.tsx [hunk 14]
```diff
@@ -48,7 +48,7 @@ const MiniToc = memo<MinitocProps>(({headings, router, headerHeight, onItemClick
 
     return (
         <nav className={b()} aria-label={t('article-navigation')}>
-            <h2 className={b('title')}>{t<string>('title')}:</h2>
+            <h2 className={b('title', 'exclusion')}>{t<string>('title')}:</h2>
```
[the remark this answers is not in the document]
Comment (JeikZim) [reply src/components/MiniToc/MiniToc.tsx]: This fixed the style error in docs-viewer-client. But it was done wrong, and now I've fixed it. Thank you

File: src/components/MobileControls/MobileControls.scss [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    @media (max-width: map-get($screenBreakpoints, 'sm') - 1) {
+        &__lang-tooltip {
+            right: 8px !important; /* stylelint-disable-line declaration-no-important */
+            left: auto !important; /* stylelint-disable-line declaration-no-important */
+            margin: 0 !important; /* stylelint-disable-line declaration-no-important */
+        }
+    }
+
+    // After 350px lang switcher doesn't fit in header
+    @media (max-width: 350px) {
```
[the remark this answers is not in the document]
Comment (JeikZim) [reply src/components/MobileControls/MobileControls.scss]: This is the code from Controls.scss, which turned out to be unnecessary. I have deleted it now.

File: src/components/MobileControls/MobileControls.tsx [hunk 16]
```diff
@@ -0,0 +1,147 @@
+/* eslint-disable react-hooks/rules-of-hooks */
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControls.tsx:1]: why is this necessary ?

File: src/components/MobileControls/MobileControls.tsx [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+type UseTranslationReturnType = TFunction<Namespace<string>, KeyPrefix<Namespace<string>>>;
+
+export interface MobileControlsProps {
+    controlSize: ControlSizes;
+    lang: Lang;
+    userSettings: DocSettings & ControlsProps;
+}
+
+const LEGACY_LANG_ITEMS = [
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControls.tsx:25]: we support quite a large number of languages for ui and any language for documentation, why are there these three ? have you checked that the documents in other languages ​​work as they should ?

File: src/components/MobileControls/MobileControls.tsx [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+];
+
+const useLangControl = (
+    t: UseTranslationReturnType,
+    lang: Lang,
+    langs?: Lang[],
+    onChangeLang?: (lang: Lang) => void,
+) => {
+    const controlName = 'lang';
+    const icon = Globe;
```
Comment (martyanovandrey) [inline src/components/MobileControls/MobileControls.tsx:38]: there is better not to allocate variables for this, check it all over the PR

File: src/components/SubNavigation/hooks/useMiniTocData.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    menuOpen: boolean,
+    // setVisibility: (event: boolean) => void,
+    // onMiniTocItemClick?: (event: MouseEvent) => void,
+) => {
+    const [miniTocOpen, setMiniTocOpen] = useState(false);
+    const [activeMiniTocTitle, setMiniTocTitle] = useState(pageTitle);
+
+    const closeMiniToc = useCallback(() => setMiniTocOpen(false), []);
+
+    // const onItemClick = (event: MouseEvent) => {
```
Comment (martyanovandrey) [inline src/components/SubNavigation/hooks/useMiniTocData.ts:15]: it is necessary to remove all commented code throughout the entire PR

File: src/components/SubNavigation/hooks/useVisibility.ts [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    );
+
+    const controlVisibility = useCallback(() => {
+        const scrollY = window.scrollY;
+
+        if (lastScrollY === null) {
+            return setLastScrollY(scrollY);
+        }
+
+        if (miniTocOpen || menuOpened || scrollY < 55) {
```
Comment (martyanovandrey) [inline src/components/SubNavigation/hooks/useVisibility.ts:20]: you need to give names to the numerical variables so that it’s clear what 55 50 etc. is

File: src/components/navigation/hooks/useItemsWithCustomDropdowns.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+    items.forEach((item) => {
+        itemsWithCustomDropdowns.push(item);
+
+        if (item.type !== 'dropdown') {
+            return;
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const newMobileItem: any = {};
```
Comment (martyanovandrey) [inline src/components/navigation/hooks/useItemsWithCustomDropdowns.ts:23]: If possible, it is better not to use any

**Review by martyanovandrey — CHANGES_REQUESTED** [review]

File: src/components/MobileControls/MobileControls.tsx (hunk 17 above)
Comment (JeikZim) [reply src/components/MobileControls/MobileControls.tsx:25]: These are the default constants that are used in the LangControl component. I just duplicated this logic. And now I'll take it out to constants.ts

Comment (JeikZim) [reply src/components/MobileControls/MobileControls.tsx:25]: These values are used only when the user has not added their languages to the .yfm-config

File: src/components/MobileControls/MobileControls.tsx (hunk 16 above)
Comment (JeikZim) [reply src/components/MobileControls/MobileControls.tsx]: This is not necessary. Now I have removed this, and refactored the logic of the component

File: src/components/navigation/hooks/useItemsWithCustomDropdowns.ts (hunk 21 above)
Comment (JeikZim) [reply src/components/navigation/hooks/useItemsWithCustomDropdowns.ts:26]: This is really necessary. And I added a comment to make it clear why 'any' is used there

Comment (JeikZim) [reply src/components/navigation/hooks/useItemsWithCustomDropdowns.ts]: UPD: you were right, you can do without it, now I have found a way, and I have already fixed it

**Review by martyanovandrey — APPROVED** [review]
